### DISCUSSION TITLE
Better way to convert to numpy array?

### DISCUSSION THREAD

**Discussion by Abhijit-2592** [DIPlib/diplib#175] [General] [answered]
Currently to convert a PyDIP image to numpy array I use the following

```
import numpy as np
c,r = dip_image.Sizes()
numpy_image = np.frombuffer(dip_image,dtype=np.uint8).reshape(r,c)
#This works for Monochrome Images, If porting color images to numpy I use reshape(r,c,3).
```
Is there a better way to do it?
Also does PyDIP use x,y instead of row,cols for indexing?

**Comment by crisluengo** [comment, marked as answer]
You can cast a `dip.Image` to `np.array` using the `np.asarray` function, and the reverse using the `dip.Image` constructor. But you don't necessarily need to do this. You can use a `dip.Image` as input to any Numpy function, and any `np.array` as input to any PyDIP function.

```
>>> a=dip.ImageReadICS('trui')
>>> import numpy as np
>>> b=np.asarray(a)
>>> b
array([[115, 113, 115, ..., 145, 145, 141],
       [119, 116, 117, ..., 147, 140, 138],
       [116, 117, 120, ..., 146, 140, 142],
       ..., 
       [ 90,  90,  91, ..., 198, 204, 203],
       [ 88,  87,  87, ..., 198, 201, 202],
       [ 88,  87,  89, ..., 185, 194, 203]], dtype=uint8)
>>> np.sum(b)
9023332
>>> np.sum(a)
9023332
```

Note that casting a Numpy array to `dip.Image` causes the two objects to share data!
